### DISCUSSION TITLE
Best way to deserialize a list with generic argument factory

### DISCUSSION THREAD

**Discussion by vware** [rrousselGit/freezed#1138] [Q&A] [answered]
I'm using a generic class to wrap my API calls, which is working really well for int, bool, etc. I needed to extend it to also be able to deserialize a list like List<String>. Currently, my setup looks like this, and it is working fine:

Outer class:
```
@Freezed(genericArgumentFactories: true)
class ApiResponse<T> with _$ApiResponse<T> {
  const factory ApiResponse({
    required T? data,
    ...
  }) = _ApiResponse<T>;

  factory ApiResponse.fromJson(
          Map<String, dynamic> json, T Function(Object? json) fromJsonT) =>
      _$ApiResponseFromJson(json, fromJsonT);
}
```

Inner class:
```
class UserNotification with _$UserNotification {
  const factory UserNotification({
    required int id,
    required int timestamp,
    ...
  }) = _UserNotification;

  factory UserNotification.fromJson(Map<String, dynamic> json) =>
      _$UserNotificationFromJson(json);
}
```

And to get it all deserialized:

```
  Future<List<UserNotification>> build() async {
    ...
    return ApiResponse.fromJson(
          response,
          (json) => (json as List<dynamic>)
              .map((e) => UserNotification.fromJson(e as Map<String, dynamic>))
              .toList(),
        ).data;
  }
```

For simple types like bool, I could just write:
`    return ApiResponse.fromJson(response, (s) => s as int?);`

But (kind of understandably), if I do something similar with a List<T>, I get an error during runtime, because a List<dynamic> can't be cast to a List<UserNotification> ...

Is there a better way to do this?

**Comment by rrousselGit** [comment, marked as answer]
Do:

```dart
final list = json as List<Object?>;

return list.map(UserNotification.fromJson).toList();
```


There's little value in that `ApiResponse` class if you don't return it anyway

**Comment by vware** [comment]
I do return it in most instances, but this had me puzzled ... thanks for the help!
